### DISCUSSION TITLE
How to propagte an error from update function

### DISCUSSION THREAD

**Discussion by d-mceneaney** [hecrj/iced#963] [Q&A] [answered]
Hi,
I am completing some logic in the update function. I would normally propagate an error from a function using the return signature below. The default update function in Iced does not seem to like this and complains.

Is it not possible to propagate an error from the update function or am I doing something wrong?

Thanks.

```rust
    fn update(&mut self, message: Message) -> Result<(), Box<dyn Error>> { //error expected trait, found enum Error, not a trait
        match message {
            Message::InputButtonPressed => {
                let path = FileDialog::new()
                    .set_location("~/")
                    .add_filter("File", &["txt"])
                    .show_open_single_file()?; //error propagated here
            }
        }
        Ok(())
    }
```

**Comment by 90-008** [comment, marked as answer]
You can't return a result from the update function, but what you can do is store the error as part of your application state. So something like:
```rust
struct App {
	last_result: Result<(), Error>,
}

impl Sandbox for App {
	fn update(&mut self, message: Message) {
		match message {
			Message::Something => {
				let result = do_something();
				if result.is_err() {
					self.last_result = result;
				}
			}
		}
	}
}
```

**Reply by d-mceneaney** [reply]
Thank you yusdacra. Very interesting. I'll have a play around with that syntax.

**Reply by Levitanus** [reply]
The only obvious solution and looks quite like a crutch. So many things could happen in update, and it would be much more clear just have some direct feedback to user and shutdown.

**Reply by ILikeTeaALot** [reply]
Would adding an Error variant to the Message enum be a decent option? i.e.

```rust
struct App {
    // ...
}

enum Message {
    Something,
    SomethingElse(f64),
    Error(AppError),
}

enum AppError {
    ErrorOne(..),
    ErrorTwo(..),
    // ... and so on
}

impl App {
    fn update(&mut self, message: Message) -> Task<Message> {
        match message {
            // ...
            Message::Error(err) => {
                  // Do something with the error
            }
        }
    }
}
```

Correction: This wouldn't work for all errors, because they would have to be `Clone` and `Send`. But for most error types that is possible.
